File: client/utils/migrate.py
#python file to extract and migrate important customer informations
# covers: items, sales history and gift cards

import pandas
import os
from datetime import datetime
import requests
#from requests import Request, Session
from tqdm import tqdm

#items: 
# Set up your Square API access token
access_token = 'EAAAlnDI3enkFLK0vaVLsFnlZAwi5K2aqAqnrMG_d_vBzyGR13Rh04Ik8lNSH9Py'
old_access_token = ''
key = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySW5mb3JtYXRpb24iOnsiaWQiOiJmNjhjNmRmZi1mOGRmLTQzNzUtYjA5Ny1mMTNmNDk0OTk3ODIiLCJlbWFpbCI6ImhiYXJpbDFAaWNsb3VkLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJwaW5fcG9saWN5Ijp7InJlZ2lvbnMiOlt7ImlkIjoiRlJBMSIsImRlc2lyZWRSZXBsaWNhdGlvbkNvdW50IjoxfSx7ImlkIjoiTllDMSIsImRlc2lyZWRSZXBsaWNhdGlvbkNvdW50IjoxfV0sInZlcnNpb24iOjF9LCJtZmFfZW5hYmxlZCI6ZmFsc2UsInN0YXR1cyI6IkFDVElWRSJ9LCJhdXRoZW50aWNhdGlvblR5cGUiOiJzY29wZWRLZXkiLCJzY29wZWRLZXlLZXkiOiI2ODFmYTNmZThmY2JmZTI5OTJmZSIsInNjb3BlZEtleVNlY3JldCI6IjcxOGRhMWFjMTRkZmNmMjVjMzM2YmZlYTI0MWUzODU2M2U1ZjJjOWNjOGJkNzdiY2RlMWE1OTY4YWQ4ZWJmNmEiLCJpYXQiOjE2ODUyODk0NDZ9.dheuwiicVcI3mM7yMo9voga4Bis7nDu7g5TJocC_xkc"
headers = {
    'Square-Version': '2023-07-26',
    'Authorization': f'Bearer {access_token}',
    'Content-Type': 'application/json'
}

# Base URL for Square API
base_url = 'https://connect.squareup.com/v2'

# Create a folder to save images
#image_folder = 'item_images'
#os.makedirs(image_folder, exist_ok=True)

def fetch_items():
    items_url = f'{base_url}/catalog/list'
    items = [];
    images_url = f'{base_url}/catalog/list?types=image'
    images = [];
    cursor = None

    while True:
        params = {'cursor': cursor} if cursor else {}
        response1 = requests.get(items_url, headers=headers, params=params)
        response2 = requests.get(images_url, headers=headers, params=params)
        response_data1 = response1.json()
        response_data2 = response2.json()
        
        items.extend(response_data1['objects'])
        images.extend(response_data2['objects'])
        cursor = response_data1.get('cursor')
        
        if not cursor:
            break
    
    return items, images

''' #print(file)
    
    ------WebKitFormBoundarymaRyYrhuiH8dFi83
Content-Disposition: form-data; name="file"; filename="s632364203791441578_p1289_i1_w236.png"
Content-Type: image/png


------WebKitFormBoundarymaRyYrhuiH8dFi83
Content-Disposition: form-data; name="pinataMetadata"

{"name":"test","keyvalues":{"description":"test description","tag":"Bouteilles & thermos"}}
------WebKitFormBoundarymaRyYrhuiH8dFi83
Content-Disposition: form-data; name="pinataOptions"

{"cidVersion":0}
------WebKitFormBoundarymaRyYrhuiH8dFi83--
    
    #--6a31b283c314f2bb4c1bbcb94feb6b1e\r\nContent-Disposition: form-data; name="pinataMetadata"; filename="pinataMetadata"\r\n\r\n{\n  "name": "Casse-t\xc3\xaate "Mushroom" - Cavallini & Co."\n}\r\n--6a31b283c314f2bb4c1bbcb94feb6b1e\r\nContent-Disposition: form-data; name="pinataOptions"; filename="pinataOptions"\r\n\r\n{\n  "cidVersion": "0"\n}\r\n
    payload = '--6a31b283c314f2bb4c1bbcb94feb6b1e\r\nContent-Disposition: form-data; name="file";' +  str(file) + '\r\n--6a31b283c314f2bb4c1bbcb94feb6b1e--\r\n'
    #request = Request('POST', url,headers,files = data).prepare()
    "-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"file\"\r\n\r\nreadstream\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"pinataMetadata\"\r\n\r\n{\n  \"name\": \"test.png\"\n}\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"pinataOptions\"\r\n\r\n{\n  \"cidVersion\": 1\n}\r\n-----011000010111000001101001--\r\n\r\n"
    #print(payload)

    #s = Session()
    #response = s.send(request)
    '''

def list_new_image(file, name): 
    url = "https://api.pinata.cloud/pinning/pinFileToIPFS"
    data = {'file': file,}
    headers = { # "Content-Type": 'multipart/form-data' #; boundary=------WebKitFormBoundarymaRyYrhuiH8dFi83
        "Authorization": key,
    }

    response = requests.post( url, files=data, headers=headers)
    return response.json()

def fetch_gift_cards():
    old_headers = {
        'Square-Version': '2023-07-26',
        'Authorization': f'Bearer {old_access_token}',
        'Content-Type': 'application/json'
    }
    gift_cards_url = f'{base_url}/gift-cards'
    gift_cards = []
    cursor = None

    while True:
        params = {'cursor': cursor} if cursor else {}
        response = requests.get(gift_cards_url, headers=old_headers, params=params)
        response_data = response.json()

        gift_cards.extend(response_data['gift_cards'])
        cursor = response_data.get('cursor')

        if not cursor:
            break

    return gift_cards

# Fetch and save gift card data


def create_gift_card(type, balance, gan):
    gift_card_url = f'{base_url}/gift-cards'
    payload = {
        'idempotency_key': gan,  # Using GAN as idempotency key for uniqueness
        'type': type,
        'gan': gan,
        'balance_money': {
            'amount': balance,
            'currency': 'CAD'
        }
    }

    response = requests.post(gift_card_url, headers=headers, json=payload)
    return response.json()

def load_items_csv():

    #important categories: 
    """ 1: Item Name
        2: Description
        3: Category
        4: Price
        5: Current Quantity Ma Maison Rose
        6: image : https://139611306.cdn6.editmysite.com/uploads/1/3/9/6/139611306/ + image id
    """
    
    #df = pandas.read_csv('items.csv')
    #print(df['Item Name'])
    items, images = fetch_items()
    #print(items[0])
    #print(images)
    names = []
    descriptions = []
    categories = []
    category_index = {}
    prices = []
    scores = [] #stock or quantity

    responses = []

    #deal with items
    print("DEBUG: Processing Categories")
    for item in tqdm(items):
        category_data = item.get('category_data')
        if category_data:
            if category_data['name'] not in category_index:
                category_index[item['id']] = category_data['name']
   

    print("DEBUG: Processing Items")
    for item in tqdm(items):
        item_data = item.get('item_data', {})
        if 'name' in item_data:
            names.append(item_data["name"])
            response = requests.get(f"https://connect.squareup.com/v2/inventory/{item_data['variations'][0]['id']}", headers=headers)
            res = response.json()
            scores.append(int(res['counts'][0]["quantity"]))
        if 'description' in item_data:
            descriptions.append(item_data['description'])
        if 'category_id' in item_data:
            categories.append(category_index[item_data['category_id']])
        if 'variations' in item_data:
            prices.append(item_data['variations'][0]['item_variation_data']['price_money']['amount']/100)
    

    

        

    #deal with images
    print("DEBUG: Processing Images")
    for image in tqdm(images):
        image_data = image.get('image_data', {})
        if 'url' in image_data:
            image_url = image_data['url']

            #load image
            response = requests.get(image_url)

            #pin it 
            ipfsHash = list_new_image(response.content, '')

            #mint it to block chain
            mint_url = 'https://f5auzuxklj.execute-api.ca-central-1.amazonaws.com/dev/oracleMint'
            price = prices[images.index(image)]
            fee =  float(price *0.029 + 4.6)
            body1 = {
                "address": "0x3190b9754f22dd2b0514feff6bd299ee7514c777",
                "uri": "https://ipfs.io/ipfs/" + ipfsHash['IpfsHash'],
                "MaxPrice": float(f'{float(price - fee):.2f}'),
                "numDays": 10,
                "mintingAddress": "0x666f393A06285c3Ec10895D4092d9Dc86aeFD45b",
                "ddsAddress": "0xa244B3e1e6Bd2ccf1D226F3E269D0Af88Ef86CEE",
            }
         
            response_mint = requests.post(mint_url, json=body1)
            #print(response_mint.text)
            response_mint = response_mint.json()

            #update the cloud
            cloud_url = 'https://f5auzuxklj.execute-api.ca-central-1.amazonaws.com/dev/listItem'
            body2 = {
                "address": "0x3190b9754f22dd2b0514feff6bd299ee7514c777",
                "itemid": int(response_mint['hex'], 16),
                "name": names[images.index(image)],
                "score": scores[images.index(image)],
                "tag": categories[images.index(image)],
                "price": int(float(f'{float(price - fee):.2f}') *100000), #parseInt((itemPrice - itemFee).toFixed(2)*100000) 
                "description": descriptions[images.index(image)],
                "image": "https://ipfs.io/ipfs/" + ipfsHash['IpfsHash']
            }
            response_cloud = requests.post(cloud_url, json=body2)
            responses.append(response_cloud.text)
    print("DEBUG logs: ")
    print(responses)
    print("DEBUG: Finished (code 0)")
            

def load_sales_data():
    #get both number of sales by month and value of does sales
    data_last_year = [0] * 12
    money_last_year = [0] * 12
    #start_date = '2023-08-01T19:34:33.524Z'
    #end_date = '2024-08-01T19:34:33.524Z'
  

           
    old_headers = {
            'Square-Version': '2023-07-26',
            'Authorization': f'Bearer {old_access_token}',
            'Content-Type': 'application/json'
    }
    
    sales = []
    cursor = None
    while True:
        
        payment_url = f'{base_url}/payments?cursor={cursor}&begin_time=2023-08-01T19%3A34%3A33.524Z&end_time=2024-08-01T19%3A34%3A33.524Z'
        response = requests.get(payment_url, headers=old_headers, params={})
        response_data = response.json()
        sales.extend(response_data['payments'])
        cursor = response_data.get('cursor')

        if not cursor:
            break

    for sale in tqdm(sales):
    
        date = datetime.fromisoformat(sale['updated_at'])
        month = date.month
        data_last_year[month-1] += 1
        money_last_year[month-1] += sale['amount_money']['amount']

        '''
        var date = new Date(Date.parse(list_of_buying_ip_transac[i]));
                    let month = date.getMonth()
                    let index = labels_index[month-1]
                    let position = labels.indexOf(index)
                    data.datasets[1].data[position] +=1
                    data2.datasets[0].data[position] += list_of_buying_ip_transac_value[i]'''


    
  

           

#read the csv, gets the properties and upload them using the real upload item (using both aws and ipfs for file storage)


#sales data:

#gift cards / customers


if __name__ == '__main__':
    #deal with items
    #print(int('0x14', 16))
    #load_items_csv()

    #deal with gift cards
    print("DEBUG: Starting gift cards")
    #gift_cards = fetch_gift_cards()
    #if this doesnt work: https://squareup.com/help/us/en/article/6355-send-an-egift-card
    #for card in tqdm(gift_cards):
    #    res = create_gift_card(card['type'], card['balance_money']['amount'], card['gan'])

    #deal with sales data
    load_sales_data()

File: client/utils/migrate.js
const axios = require('axios');
const { DateTime } = require('luxon'); // You can use Luxon for better date handling

const oldAccessToken = ''; // Replace with the actual old access token
const oldHeaders = {
    'Square-Version': '2023-07-26',
    'Authorization': `Bearer ${oldAccessToken}`,
    'Content-Type': 'application/json',
};

const accessToken = 'EAAAlnDI3enkFLK0vaVLsFnlZAwi5K2aqAqnrMG_d_vBzyGR13Rh04Ik8lNSH9Py';
const headers = {
    'Square-Version': '2023-07-26',
    'Authorization': `Bearer ${accessToken}`,
    'Content-Type': 'application/json',
};

const baseUrl = 'https://connect.squareup.com/v2';
const key = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySW5mb3JtYXRpb24iOnsiaWQiOiJmNjhjNmRmZi1mOGRmLTQzNzUtYjA5Ny1mMTNmNDk0OTk3ODIiLCJlbWFpbCI6ImhiYXJpbDFAaWNsb3VkLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJwaW5fcG9saWN5Ijp7InJlZ2lvbnMiOlt7ImlkIjoiRlJBMSIsImRlc2lyZWRSZXBsaWNhdGlvbkNvdW50IjoxfSx7ImlkIjoiTllDMSIsImRlc2lyZWRSZXBsaWNhdGlvbkNvdW50IjoxfV0sInZlcnNpb24iOjF9LCJtZmFfZW5hYmxlZCI6ZmFsc2UsInN0YXR1cyI6IkFDVElWRSJ9LCJhdXRoZW50aWNhdGlvblR5cGUiOiJzY29wZWRLZXkiLCJzY29wZWRLZXlLZXkiOiI2ODFmYTNmZThmY2JmZTI5OTJmZSIsInNjb3BlZEtleVNlY3JldCI6IjcxOGRhMWFjMTRkZmNmMjVjMzM2YmZlYTI0MWUzODU2M2U1ZjJjOWNjOGJkNzdiY2RlMWE1OTY4YWQ4ZWJmNmEiLCJpYXQiOjE2ODUyODk0NDZ9.dheuwiicVcI3mM7yMo9voga4Bis7nDu7g5TJocC_xkc"


//items

async function fetchItems() {
    const itemsUrl = `${baseUrl}/catalog/list`;
    const imagesUrl = `${baseUrl}/catalog/list?types=image`;
    let items = [];
    let images = [];
    let cursor = null;

    while (true) {
        try {
            const params = cursor ? { cursor: cursor } : {};
            const response1 = await axios.get(itemsUrl, { headers: headers, params: params });
            const response2 = await axios.get(imagesUrl, { headers: headers, params: params });

            items = [...items, ...response1.data.objects];
            images = [...images, ...response2.data.objects];

            cursor = response1.data.cursor;
            if (!cursor) break;
        } catch (error) {
            console.error("Error fetching items:", error);
            break;
        }
    }
    return { items, images };
}

async function listNewImage(file, name) {
    const url = "https://api.pinata.cloud/pinning/pinFileToIPFS";
    const form = new FormData();
    
    // If the file is a file path
    form.append('file', file); //fs.createReadStream(filePath)
    
    // Alternatively, if `file` is a buffer or a file object directly, you can do:
    // form.append('file', file);

    const headers = {
        ...form.getHeaders(),
        "Authorization": key, // Replace `key` with your actual authorization key
    };

    try {
        const response = await axios.post(url, form, { headers: headers });
        return response.data;
    } catch (error) {
        console.error("Error uploading file to IPFS:", error);
        throw error;
    }
}

async function loadItems() {
    console.log("DEBUG: Fetching Items and Images");
    const { items, images } = await fetchItems(); // Assuming `fetchItems` is an async function in Node.js

    let names = [];
    let descriptions = [];
    let categories = [];
    let categoryIndex = {};
    let prices = [];
    let scores = [];
    let responses = [];

    console.log("DEBUG: Processing Categories");
    items.forEach(item => {
        const categoryData = item.category_data;
        if (categoryData) {
            if (!categoryIndex[categoryData.name]) {
                categoryIndex[item.id] = categoryData.name;
            }
        }
    });

    console.log("DEBUG: Processing Items");
    for (const item of items) {
        const itemData = item.item_data || {};
        if (itemData.name) {
            names.push(itemData.name);
            try {
                const response = await axios.get(`https://connect.squareup.com/v2/inventory/${itemData.variations[0].id}`, { headers: headers });
                const res = response.data;
                scores.push(parseInt(res.counts[0].quantity, 10));
            } catch (error) {
                console.error("Error fetching inventory:", error);
            }
        }
        if (itemData.description) {
            descriptions.push(itemData.description);
        }
        if (itemData.category_id) {
            categories.push(categoryIndex[itemData.category_id]);
        }
        if (itemData.variations) {
            prices.push(itemData.variations[0].item_variation_data.price_money.amount / 100);
        }
    }

    console.log("DEBUG: Processing Images");
    for (const image of images) {
        const imageData = image.image_data || {};
        if (imageData.url) {
            try {
                const response = await axios.get(imageData.url, { responseType: 'arraybuffer' });
                
                const ipfsHash = await listNewImage(response.data, '');

                const price = prices[images.indexOf(image)];
                const fee = parseFloat((price * 0.029 + 4.6).toFixed(2));

                const mintUrl = 'https://f5auzuxklj.execute-api.ca-central-1.amazonaws.com/dev/oracleMint';
                const body1 = {
                    address: "0x3190b9754f22dd2b0514feff6bd299ee7514c777",
                    uri: `https://ipfs.io/ipfs/${ipfsHash.IpfsHash}`,
                    MaxPrice: parseFloat((price - fee).toFixed(2)),
                    numDays: 10,
                    mintingAddress: "0x666f393A06285c3Ec10895D4092d9Dc86aeFD45b",
                    ddsAddress: "0xa244B3e1e6Bd2ccf1D226F3E269D0Af88Ef86CEE",
                };

                const responseMint = await axios.post(mintUrl, body1);
                const mintData = responseMint.data;

                const cloudUrl = 'https://f5auzuxklj.execute-api.ca-central-1.amazonaws.com/dev/listItem';
                const body2 = {
                    address: "0x3190b9754f22dd2b0514feff6bd299ee7514c777",
                    itemid: parseInt(mintData.hex, 16),
                    name: names[images.indexOf(image)],
                    score: scores[images.indexOf(image)],
                    tag: categories[images.indexOf(image)],
                    price: parseInt((price - fee).toFixed(2) * 100000),
                    description: descriptions[images.indexOf(image)],
                    image: `https://ipfs.io/ipfs/${ipfsHash.IpfsHash}`,
                };

                const responseCloud = await axios.post(cloudUrl, body2);
                responses.push(responseCloud.data);
            } catch (error) {
                console.error("Error processing image or minting:", error);
            }
        }
    }

    console.log("DEBUG logs: ", responses);
    console.log("DEBUG: Finished (code 0)");
}


//sales

async function loadSalesData() {
    let dataLastYear = Array(12).fill(0);
    let moneyLastYear = Array(12).fill(0);

    // Specify the date range (last 12 months)
    const beginTime = '2023-08-01T19:34:33.524Z';
    const endTime = '2024-08-01T19:34:33.524Z';

    let sales = [];
    let cursor = null;

    while (true) {
        try {
            const paymentUrl = `${baseUrl}/payments?cursor=${cursor}&begin_time=${beginTime}&end_time=${endTime}`;
            const response = await axios.get(paymentUrl, { headers: oldHeaders });
            const responseData = response.data;
            sales = [...sales, ...responseData.payments];
            cursor = responseData.cursor;

            if (!cursor) break;
        } catch (error) {
            console.error("Error fetching sales data:", error);
            break;
        }
    }

    sales.forEach(sale => {
        const date = DateTime.fromISO(sale.updated_at);
        const month = date.month;
        dataLastYear[month - 1] += 1;
        moneyLastYear[month - 1] += sale.amount_money.amount;
    });

    const transferparams = {
        TableName: "partnerlogin",
        Key: {
          email: req.body.email,
        },
        ExpressionAttributeNames: { '#td': 'transfer_data' },
        ExpressionAttributeValues: {},
        ReturnValues: 'UPDATED_NEW',
      };
      payparams.UpdateExpression = 'SET '
      payparams.ExpressionAttributeValues[':transferData'] = [dataLastYear, moneyLastYear];
      payparams.UpdateExpression += '#td = :transferData'
  
      dynamodb.update(transferparams, (error, result) => {
          if (error) {
            console.log(error.message);
            res.json({error: error.message, params: transferparams})
          }
          else {
            res.send("done")
          }
      });

    return { dataLastYear, moneyLastYear };
}



//giftcards

async function fetchGiftCards() {
    let giftCards = [];
    let cursor = null;

    while (true) {
        try {
            const giftCardsUrl = `${baseUrl}/gift-cards?cursor=${cursor}`;
            const response = await axios.get(giftCardsUrl, { headers: headers });
            const responseData = response.data;

            giftCards = [...giftCards, ...responseData.gift_cards];

            cursor = responseData.cursor;
            if (!cursor) break;
        } catch (error) {
            console.error("Error fetching gift cards:", error);
            break;
        }
    }

    return giftCards;
}

async function createGiftCard(type, amount, gan) {
    const url = `${baseUrl}/gift-cards`;
    const data = {
        idempotency_key: gan,
        type: type,
        balance_money: {
            amount: amount,
            currency: 'CAD' // Replace with the appropriate currency if needed
        },
        gan: gan
    };

    try {
        const response = await axios.post(url, data, { headers: headers });
        return response.data;
    } catch (error) {
        console.error("Error creating gift card:", error.response ? error.response.data : error.message);
        throw error;
    }
}



//shopify migration: 

//items
//https://shopify.dev/docs/api/admin-graphql/2024-10/queries/products?language=cURL
// Shopify credentials
const SHOPIFY_STORE_URL = 'your-store-name.myshopify.com';
const SHOPIFY_ACCESS_TOKEN = 'your-access-token';

// Base URL for Shopify API
const SHOPIFY_API_URL = `https://${SHOPIFY_STORE_URL}/admin/api/2023-10/products.json`;

// Function to fetch products from Shopify
async function fetchProducts() {
    try {
        const response = await axios.get(SHOPIFY_API_URL, {
            headers: {
                'X-Shopify-Access-Token': SHOPIFY_ACCESS_TOKEN
            }
        });
        return response.data.products;
    } catch (error) {
        console.error('Error fetching products:', error.message);
        return [];
    }
}

// Main function to fetch products and save their images
async function saveProductImages() {
    let names = [];
  let descriptions = [];
  let categories = [];
  let categoryIndex = {};
  let prices = [];
  let scores = [];
  let responses = [];
    const products = await fetchProducts();

    for (const product of products) {
        console.log(`Processing product: ${product.title}`);
        const image = product.media.nodes.preview.image.url; //use media.first

        const response = await nfetch(image, { responseType: 'arraybuffer' });
              
        const ipfsHash = await listNewImage(response.data, '');
        const name = product.title
        const score = product.totalInventory
        const description = product.description
        const category = product.tags[0]
        const price = product.priceRangeV2.maxVariantPrice.amount;
        const fee = parseFloat((price * 0.029 + 4.6).toFixed(2));

        if (!last_id) {
        //pull the last id
        }

        /* const mintUrl = 'https://f5auzuxklj.execute-api.ca-central-1.amazonaws.com/dev/oracleMint';
        const body1 = {
            address: address,
            uri: `https://ipfs.io/ipfs/${ipfsHash.IpfsHash}`,
            MaxPrice: parseFloat((price - fee).toFixed(2)),
            numDays: 10,
            mintingAddress: "0x666f393A06285c3Ec10895D4092d9Dc86aeFD45b",
            ddsAddress: "0xa244B3e1e6Bd2ccf1D226F3E269D0Af88Ef86CEE",
        };

        const responseMint = await nfetch(mintUrl, {body:body1});
        const mintData = responseMint.data;*/

        const cloudUrl = 'https://f5auzuxklj.execute-api.ca-central-1.amazonaws.com/dev/listItem';
        const body2 = {
            address: address,
            itemid: parseInt(mintData.hex, 16),
            name: name,
            score: score,
            tag: category,
            price: parseInt((price - fee).toFixed(2) * 100000),
            description: description,
            image: `https://ipfs.io/ipfs/${ipfsHash.IpfsHash}`,
        };

        const responseCloud = await nfetch(cloudUrl, {body:body2});
        responses.push(responseCloud.data);
    }

    console.log('All images have been downloaded.');
}

//sales

// Base URL for Shopify Orders API
const SHOPIFY_ORDERS_URL = `https://${SHOPIFY_STORE_URL}/admin/api/2023-10/orders.json`;

// Function to fetch orders from Shopify
async function fetchOrders(email) {
    try {
        let dataLastYear = Array(12).fill(0);
        let moneyLastYear = Array(12).fill(0);
        const response = await axios.get(SHOPIFY_ORDERS_URL, {
            headers: {
                'X-Shopify-Access-Token': SHOPIFY_ACCESS_TOKEN
            },
            params: {
                status: 'any', // Fetch orders with any status (open, closed, or cancelled)
                limit: 250     // Fetch up to 250 orders per request (API maximum)
            }
        });

        response.data.orders.forEach(sale => {
            const date = DateTime.fromISO(sale.closed_at);
            const month = date.month;
            dataLastYear[month - 1] += 1;
            moneyLastYear[month - 1] += sale.currentTotalPriceSet.presentmentMoney.amount;
        });
        dataLastYear = response.data.orders
        const transferparams = {
            TableName: "partnerlogin",
            Key: {
              email: email,
            },
            ExpressionAttributeNames: { '#td': 'transfer_data' },
            ExpressionAttributeValues: {},
            ReturnValues: 'UPDATED_NEW',
          };
          payparams.UpdateExpression = 'SET '
          payparams.ExpressionAttributeValues[':transferData'] = [dataLastYear, moneyLastYear];
          payparams.UpdateExpression += '#td = :transferData'
      
          dynamodb.update(transferparams, (error, result) => {
              if (error) {
                console.log(error.message);
                res.json({error: error.message, params: transferparams})
              }
              else {
                res.send("done")
              }
          });
  
        return { dataLastYear, moneyLastYear };
    
      
    } catch (error) {
        console.error('Error fetching orders:', error.message);
        return [];
    }
}

//gift cards

// Base URL for Shopify Gift Cards API
const SHOPIFY_GIFT_CARDS_URL = `https://${SHOPIFY_STORE_URL}/admin/api/2023-10/gift_cards.json`;

// Function to fetch gift cards from Shopify
async function fetchGiftCards() {
    try {
        const response = await axios.get(SHOPIFY_GIFT_CARDS_URL, {
            headers: {
                'X-Shopify-Access-Token': SHOPIFY_ACCESS_TOKEN
            },
            params: {
                limit: 250, // Fetch up to 250 gift cards per request
                status: 'enabled' // Fetch only enabled gift cards (optional)
            }
        });
        return response.data.gift_cards;
    } catch (error) {
        console.error('Error fetching gift cards:', error.message);
        return [];
    }
}

const NEW_SHOPIFY_STORE_URL = 'new-store-name.myshopify.com';
const NEW_SHOPIFY_ACCESS_TOKEN = 'new-access-token';

// Base URL for Shopify Gift Cards API
const SHOPIFY_GIFT_CARD_CREATE_URL = `https://${NEW_SHOPIFY_STORE_URL}/admin/api/2023-10/gift_cards.json`;

// Function to create a gift card
async function createGiftCard(giftCardData) {
    try {
        const response = await axios.post(
            SHOPIFY_GIFT_CARD_CREATE_URL,
            {
                gift_card: {
                    initial_value: giftCardData.initial_value,
                    balance: giftCardData.balance,
                    note: giftCardData.note || `Migrated from old account`,
                    expires_on: giftCardData.expires_on || null
                }
            },
            {
                headers: {
                    'X-Shopify-Access-Token': NEW_SHOPIFY_ACCESS_TOKEN,
                    'Content-Type': 'application/json'
                }
            }
        );

        console.log(`Gift card created with ID: ${response.data.gift_card.id}`);
    } catch (error) {
        console.error(`Error creating gift card: ${giftCardData.id}`, error.message);
    }
}


async function migrateGiftCard() {
    const gift_cards = await fetchGiftCards()
    for (const gift_card in gift_cards) {
        const amount = gift_card.balance
        const gan = gift_card.customer.email

        createGiftCard(amount, gan)

        
    }
}
